src/utils/src/net/macvtap.rs

```diff
@@ -0,0 +1,165 @@
+                    macvtap_dir.display()
+                ),
+            ));
+        }
+
+        // unwrap() is safe because there is one element in dirents and we already checked for that.
+        let dirent = dirents.into_iter().next().unwrap();
+        let dev_name = dirent?.file_name(); // e.g. tap42
+
+        // Read /sys/.../<if_name>/macvtap/<dev_name>/dev
+        let dev_file_path = macvtap_dir.join(&dev_name).join("dev");
+        let dev_str = std::fs::read_to_string(&dev_file_path)?;
+
+        // Trim trailing newline, split on ':', parse major and minor numbers
+        let mut cpts = dev_str.trim_end().splitn(2, ':');
+        let major = parse_dev(&mut cpts);
+        let minor = parse_dev(&mut cpts);
+
+        if major.is_none() || minor.is_none() {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                format!("Invalid {} contents: {}", dev_file_path.display(), dev_str),
+            ));
+        }
+
+        Ok(Self {
+            if_name: if_name.to_string(),
+            tap_name: dev_name.into_string().unwrap(),
+            major: major.unwrap(),
+            minor: minor.unwrap(),
+        })
+    }
+
+    /// Creates a device node for the given network MacVTap instance at the given path with
+    /// the given owner and group.
+    pub fn mknod<P: AsRef<Path>>(&self, path: P, uid: u32, gid: u32) -> Result<()> {
+        let osstr = path.as_ref().as_os_str();
+        let cstr =
+            CString::new(osstr.as_bytes()).map_err(|x| Error::new(ErrorKind::InvalidInput, x))?;
+
+        // Safety: path is a C-compatible pointer to memory that is in scope
+        SyscallReturnCode(unsafe {
+            libc::mknod(
+                cstr.as_ptr(),
+                libc::S_IFCHR | libc::S_IRUSR | libc::S_IWUSR,
+                libc::makedev(self.major, self.minor),
+            )
+        })
+        .into_empty_result()?;
+
+        // Safety: path is a C-compatible pointer to memory that is in scope
+        SyscallReturnCode(unsafe { libc::chown(cstr.as_ptr(), uid, gid) }).into_empty_result()
+    }
+}
+
+fn parse_dev<'a, It: Iterator<Item = &'a str>>(it: &mut It) -> Option<u32> {
+    it.next().and_then(|s| s.parse().ok())
+}
+
+fn is_normal_path_component(if_name: &str) -> bool {
+    // Mustn't be empty, "." or ".."; mustn't contain / or \.
+    if_name != ""
+        && if_name != "."
+        && if_name != ".."
+        && if_name.as_bytes().iter().all(|b| *b != b'/' && *b != b'\\')
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_is_normal_path_component() {
+        assert!(!is_normal_path_component(""));
+        assert!(!is_normal_path_component("."));
+        assert!(!is_normal_path_component(".."));
+        assert!(!is_normal_path_component("/"));
+        assert!(!is_normal_path_component("./x"));
+        assert!(!is_normal_path_component("../x"));
+        assert!(!is_normal_path_component("../x"));
```

Comment:
super nit: these two checks seem to be duplicated. You could remove one of them.

Reply:
Done.